### PR TITLE
[DNM] 6.5.2 debug patch: print select result

### PR DESCRIPTION
**Pull request by ekexium**
### What problem does this PR solve?

Issue Number: close #xxx

Problem Summary:

For a select statement like
```sql
create table print_result (c int primary key, v varchar(20))
insert into print_result values (1, 'a')
insert into print_result values (2, 'b')
select * from print_result
```

Print a log like
```log
[2023/08/01 12:58:31.255 +08:00] [INFO] [util.go:607] ["statement result"] [conn=4014666597707809173] [txnStartTS=443250353436884993] [forUpdateTS=443250353436884993] [sql="select * from print_result"] [rows="(\u00011\u0001a);\n (\u00012\u0001b)"]
```

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: session/session.go [hunk 1]
```diff
@@ -2236,6 +2237,49 @@ func (s *session) ExecuteStmt(ctx context.Context, stmtNode ast.StmtNode) (sqlex
...
+		zap.Uint64("start ts", vars.TxnCtx.StartTS),
+		zap.Uint64("for update ts", vars.TxnCtx.GetForUpdateTS()),
+		zap.String("rows", "("+strings.Join(resultRows, ");\n (")+")"),
+	)
+
+	copiedRs := &sqlexec.CopiedRecordSet{
+		Records:      records,
+		ResultFields: rs.Fields(),
+	}
+	rs.Close()
```
Comment (cfzjywxk) [inline session/session.go:2279]: The lifetime of `rs` is changed compared with the original version, would it introduce extra risks I'm not sure if there are implicit dependencies of it.

Or is it possible to print the results of the `Next` call so it's unnecessary to copy or clone here?

File: session/session.go [hunk 2]
```diff
@@ -2236,6 +2237,49 @@ func (s *session) ExecuteStmt(ctx context.Context, stmtNode ast.StmtNode) (sqlex
...
+		for i := 0; i < chk.NumRows(); i++ {
+			row := chk.GetRow(i)
+			records = append(records, row)
+			resultRows = append(resultRows, row.ToString(fieldTypes))
+		}
+		chk = chunk.Renew(chk, 32)
+	}
+
+	// Print all rows
+	logutil.Logger(ctx).Info("statement returned result",
```
Comment (cfzjywxk) [inline session/session.go:2267]: If there are quite a lot of rows, there could be memory risk.

File: session/session.go [hunk 3]
```diff
@@ -2359,8 +2403,18 @@ func runStmt(ctx context.Context, se *session, s sqlexec.Statement) (rs sqlexec.
 				}
 			}
 		}
+
+		copiedRs := rs
```
Comment (cfzjywxk) [inline session/session.go:2407]: May need a switch to turn it on or off the hook.

**Review by MyonKeminta — COMMENTED** [review]
LGTM. The only problem of me is that, as the timing of performing physical read is changed (if I understood it correctly), I'm not sure whether it will affect reproduction of the issue.

File: session/session.go (hunk 2 above)
Comment (MyonKeminta) [reply session/session.go:2267]: Can we assume that the transaction sizes are not too big in the environment that we'are going to use this?
Another way to print them is to collect them when the `resultSet` is being consumed by calling `Next`

File: session/session.go (hunk 3 above)
Comment (ekexium) [reply session/session.go]: I've set the maximum number of rows to store and log to 5. Do you think it still necessary to add a switch?

Comment (cfzjywxk) [reply session/session.go]: Fine with me, the limit `5` is a little bit small, maybe `50`. If it's enough to cover the questionable query that's OK.

**Comment by ekexium** [comment]
I've changed it to a wrapper style. PTAL @MyonKeminta

File: server/util.go [hunk 4]
```diff
@@ -543,3 +549,71 @@ func (h CorsHandler) ServeHTTP(w http.ResponseWriter, req *http.Request) {
...
+	zapFields     []zap.Field
+	maxCachedRows int
+}
+
+func (r *recordSetLoggerWrapper) Fields() []*ast.ResultField {
+	return r.rs.Fields()
+}
+
+func (r *recordSetLoggerWrapper) Next(ctx context.Context, req *chunk.Chunk) error {
+	chk := r.rs.NewChunk(nil)
```
Comment (MyonKeminta) [inline server/util.go:576]: It looks to me that the better way is to directly read to `req` and then get some rows from it, but it seems kind of difficult to know when `req` is reset? 🤔

**Review by MyonKeminta — COMMENTED** [review]
LGTM

File: server/util.go (hunk 4 above)
Comment (ekexium) [reply server/util.go:576]: Yeah that's not something we can control.

**Pull request closed without merging**